pkg/cloudprovider/aws/fleet/packing/packing.go

```
@@ -16,41 +16,187 @@ package packing
-	return &PodPacker{ec2: ec2}
+// NewPacker returns a Packer implementation
+func NewPacker(ec2 ec2iface.EC2API) Packer {
+	return &podPacker{ec2: ec2}
 }
 
 // Pack returns the packings for the provided pods. Computes a set of viable
 // instance types for each packing of pods. Instance variety enables EC2 to make
-// better cost and availability decisions.
-func (p *PodPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
-	zap.S().Debugf("Successfully packed %d pod(s) onto %d node(s)", len(pods), 1)
-	return []*Packings{
-		{
-			InstanceTypeOptions: []string{"m5.large"}, // TODO, prioritize possible instance types
-			Pods:                pods,
-		},
-	}, nil
+// better cost and availability decisions. Pods provided are all schedulable in
+// the same zone as tightly as possible. It follows the First Fit Decreasing bin
+// packing technique, reference-
+// https://en.wikipedia.org/wiki/Bin_packing_problem#First_Fit_Decreasing_(FFD)
+func (p *podPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packing, error) {
+	// TODO use ctx when calling ec2 API
+	return p.packPods(pods)
+}
+
+// takes a list of pods, sorts them based on their resource requirements compared by CPU and memory.
+func (p *podPacker) packPods(pods []*v1.Pod) ([]*Packing, error) {
+	// Sort pods in decreasing order by the amount of CPU requested, if
+	// CPU requested is equal compare memory requested.
+	sort.Sort(sort.Reverse(byResourceRequested{pods}))
+	estimator := newPackingEstimator()
+	packings := []*Packing{}
+	// start with the biggest pod check max how many pods can we fit with this pod
+	for start, pod := range pods {
+		if estimator.isPodPacked(pod) {
+			continue
+		}
+		packing := estimator.calculatePackingForPod(start, pods)
+		// checked all instance type and found no packing option
+		if len(packing.Pods) == 0 || len(packing.InstanceTypes) == 0 {
+			return nil, fmt.Errorf("no instance type found for packing %d pods", len(pods))
```

Should we return an error, or just log a debug line? Maybe we should check if len(packings) == 0 and then error?

---

I think here it should return an error because if we are not able to pack a single pod, essentially because there was no right instance type with right capacity available, it should fail because bin packing failed.
`packing` is a struct not sure if you meant checking `len(packings)` at a higher function level?